keras_retinanet/utils/transform.py

```diff
@@ -0,0 +1,241 @@
+    The new AABB contains all corner points of the original AABB after applying the given transformation.
+
+    # Arguments
+        transform: The transormation to apply.
+        x1:        The minimum X value of the AABB.
+        y1:        The minimum y value of the AABB.
+        x2:        The maximum X value of the AABB.
+        y2:        The maximum y value of the AABB.
+    # Returns
+        The new AABB as tuple (x1, y1, x2, y2)
+    """
+    x1, y1, x2, y2 = aabb
+    # Transform all 4 corners of the AABB.
+    points = transform.dot([
+        [x1, x2, x1, x2],
+        [y1, y2, y2, y1],
+        [1,  1,  1,  1 ],
+    ])
+
+    # Extract the min and max corners again.
+    min_corner = points.min(axis=1)
+    max_corner = points.max(axis=1)
+
+    return [min_corner[0], min_corner[1], max_corner[0], max_corner[1]]
+
+
+def _random_vector(min, max, prng = DEFAULT_PRNG):
+    """ Construct a random vector between min and max.
+    # Arguments
+        min: the minimum value for each component
+        max: the maximum value for each component
+    """
+    min = np.array(min)
+    max = np.array(max)
+    assert min.shape == max.shape
+    assert len(min.shape) == 1
+    return prng.uniform(min, max)
+
+
+def rotation(angle):
+    """ Construct a homogeneous 2D rotation matrix.
+    # Arguments
+        angle: the angle in radians
+    # Returns
+        the rotation matrix as 3 by 3 numpy array
+    """
+    return np.array([
+        [np.cos(angle), -np.sin(angle), 0],
+        [np.sin(angle),  np.cos(angle), 0],
+        [0, 0, 1]
+    ])
+
+
+def random_rotation(min, max, prng = DEFAULT_PRNG):
+    """ Construct a random rotation between -max and max.
+    # Arguments
+        min:  a scalar for the minumum absolute angle in radians
+        max:  a scalar for the maximum absolute angle in radians
+        prng: the pseudo-random number generator to use.
+    # Returns
+        a homogeneous 3 by 3 rotation matrix
+    """
+    return rotation(prng.uniform(min, max))
+
+
+def translation(translation):
+    """ Construct a homogeneous 2D translation matrix.
+    # Arguments
+        translation: the translation 2D vector
+    # Returns
+        the translation matrix as 3 by 3 numpy array
+    """
+    return np.array([
+        [1, 0, translation[0]],
+        [0, 1, translation[1]],
+        [0, 0, 1]
+    ])
+
+
+def random_translation(min, max, prng = DEFAULT_PRNG):
+    """ Construct a random 2D translation between min and max.
+    # Arguments
+        min:  a 2D vector with the minumum translation for each dimension
+        max:  a 2D vector with the maximum translation for each dimension
+        prng: the pseudo-random number generator to use.
+    # Returns
+        a homogeneous 3 by 3 translation matrix
+    """
+    return translation(_random_vector(min, max, prng))
+
+
+def shear(amount):
+    """ Construct a homogeneous 2D shear matrix.
+    # Arguments
+        amount: the shear amount
```

Comment:
Maybe make it a bit more clear what it is. Presumably it is the angle of the shear in radians?

Reply:
It is an angle, though the matrix generated isn't actually a shear matrix. However, I copied it from `keras` since I figured we should probably use the same definition of shear as `keras` does, even if it is technically incorrect.

I believe the `keras` definition of shear compensates for the stretching that a real shear operation would do by reducing the Y component as the X component gets larger. So it's sort of halfway between a shear and a rotation.

Another thing I wasn't really happy about is that this shear functions only shears parallel to the X axis. It's not very generic.